### PR TITLE
feat: richtext Comments

### PR DESCRIPTION
**Pull request by DarkPhoenix2704**
## Change Summary

Provide summary of changes with issue number if any.

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Provide summary of changes.

## Additional information / screenshots (optional)

Anything for maintainers to be made aware of

### CODE REVIEW

File: packages/nc-gui/components/smartsheet/expanded-form/index.vue [hunk 1]
```diff
@@ -129,7 +129,8 @@ const {
   saveRowAndStay,
   row: _row,
   save: _save,
-  loadCommentsAndLogs,
+  loadComments,
+  loadAudits,
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/components/smartsheet/expanded-form/index.vue:133]: loadComments and loadAudits are different functions. This comment is irrelevant

File: packages/nc-gui/components/smartsheet/expanded-form/index.vue [hunk 2]
```diff
@@ -320,13 +321,13 @@
...
     await _loadRow()
-    await loadCommentsAndLogs()
+    await Promise.all([loadComments(), loadAudits()])
   }
 
   if (props.rowId) {
     try {
       await _loadRow(props.rowId)
-      await loadCommentsAndLogs()
+      await Promise.all([loadComments(), loadAudits()])
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/components/smartsheet/expanded-form/index.vue:330]: Error are handled in the respective functions in useExpandedFormStore. 
comment is irrelevant

File: packages/nc-gui/components/smartsheet/expanded-form/Comments.vue [hunk 3]
```diff
@@ -123,7 +121,10 @@
 
   try {
     await _saveComment()
-
+    await nextTick(() => {
+      commentInputRef?.value?.setEditorContent('', true)
+      isExpandedFormCommentMode.value = true
+    })
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/components/smartsheet/expanded-form/Comments.vue:125]: This is handled in the function itself. 
COmment is irrelevant

File: packages/nocodb/src/utils/acl.ts [hunk 4]
```diff
@@ -210,6 +211,7 @@ const rolePermissions:
       commentsCount: true,
       commentRow: true,
       commentUpdate: true,
+      commentDelete: true,
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/utils/acl.ts:214]: ignore. Not associated with this pr

File: packages/nc-gui/components/smartsheet/expanded-form/Comments.vue [hunk 5]
```diff
@@ -1,18 +1,30 @@
...
+} = useExpandedFormStoreOrThrow()
 
 const { isExpandedFormCommentMode } = storeToRefs(useConfigStore())
 
 const commentsWrapperEl = ref<HTMLDivElement>()
 
+const commentInputRef = ref<any>()
+
+const editRef = ref<any>()
+
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/components/smartsheet/expanded-form/Comments.vue:25]: ignore

File: packages/nocodb/src/models/Model.ts [hunk 6]
```diff
@@ -433,7 +432,7 @@ export default class Model implements TableType {
   }
 
   async delete(ncMeta = Noco.ncMeta, force = false): Promise<boolean> {
-    await Audit.deleteRowComments(this.id, ncMeta);
+    await Comment.deleteRowComments(this.id, ncMeta);
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/models/Model.ts:435]: they are handled propely

File: packages/nc-gui/components/smartsheet/expanded-form/RichTextOptions.vue [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+      <template #title>
+        <div class="flex flex-col items-center">
+          <div>
+            {{ $t('labels.bold') }}
+          </div>
+          <div class="text-xs">{{ cmdOrCtrlKey }} B</div>
+        </div>
+      </template>
+      <NcButton
+        :class="{ 'is-active': editor?.isActive('bold') }"
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/components/smartsheet/expanded-form/RichTextOptions.vue:86]: ignore

File: packages/nc-gui/components/smartsheet/expanded-form/RichTextOptions.vue [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+  if (isActiveNodeMarkActive) {
+    editor.value!.chain().focus().unsetLink().run()
+  } else {
+    if (editor.value.state.selection.empty) {
+      editor
+        .value!.chain()
+        .focus()
+        .insertContent(' ')
+        .setTextSelection({ from: editor.value!.state.selection.$from.pos, to: editor.value!.state.selection.$from.pos + 1 })
+        .toggleLink({
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/components/smartsheet/expanded-form/RichTextOptions.vue:42]: ignore

File: packages/nocodb/src/meta/meta.service.ts [hunk 9]
```diff
@@ -252,6 +252,15 @@ export class MetaService {
...
         break;
+      case MetaTable.COMMENTS:
+        prefix = 'com';
+        break;
+      case MetaTable.COMMENTS_REACTIONS:
+        prefix = 'cre';
+        break;
+      case MetaTable.USER_COMMENTS_NOTIFICATIONS_PREFERENCE:
+        prefix = 'cnp';
+        break;
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/meta/meta.service.ts:263]: ignore

File: packages/nocodb/src/meta/migrations/XcMigrationSourcev2.ts [hunk 10]
```diff
@@ -32,6 +32,7 @@ import * as nc_042_user_block from '~/meta/migrations/v2/nc_042_user_block';
 import * as nc_043_user_refresh_token from '~/meta/migrations/v2/nc_043_user_refresh_token';
 import * as nc_044_view_column_index from '~/meta/migrations/v2/nc_044_view_column_index';
 import * as nc_045_extensions from '~/meta/migrations/v2/nc_045_extensions';
+import * as nc_046_comment_mentions from '~/meta/migrations/v2/nc_046_comment_mentions';
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/meta/migrations/XcMigrationSourcev2.ts:35]: handled

File: packages/nc-gui/composables/useExpandedFormStore.ts [hunk 11]
```diff
@@ -87,24 +101,73 @@ const [useProvideExpandedFormStore, useExpandedFormStore] = useInjectionState((m
...
+  const deleteComment = async (commentId: string) => {
+    if (!isUIAllowed('commentDelete')) return
+
+    try {
+      await $api.utils.commentDelete(commentId)
+      await loadComments()
+    } catch (e: any) {
+      message.error(e.message)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useExpandedFormStore.ts:148]: ignore. It is updated in try block itself

File: packages/nc-gui/composables/useExpandedFormStore.ts [hunk 12]
```diff
@@ -126,15 +189,15 @@
...
         fk_model_id: meta.value?.id as string,
         row_id: rowId,
-        description: `The following comment has been created: ${comment.value}`,
+        comment: `${comment.value}`,
       })
 
       reloadTrigger?.trigger()
 
-      await loadCommentsAndLogs()
+      await Promise.all([loadComments(), loadAudits()])
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useExpandedFormStore.ts:200]: ignore. They were separated

File: packages/nc-gui/composables/useExpandedFormStore.ts [hunk 13]
```diff
@@ -87,24 +101,73 @@
...
+          await $api.utils.auditList({
+            row_id: rowId,
+            fk_model_id: meta.value.id as string,
+          })
+        ).list?.reverse?.() || []
+    } catch (e: any) {
+      message.error(e.message)
+    } finally {
+      isAuditLoading.value = false
+    }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useExpandedFormStore.ts:170]: ignore

File: packages/nc-gui/composables/useExpandedFormStore.ts [hunk 14]
```diff
@@ -87,24 +101,73 @@
...
+          ...comment,
+          created_display_name: user?.display_name ?? (user?.email ?? '').split('@')[0],
+        }
+      })
+    } catch (e: any) {
+      message.error(e.message)
+    } finally {
+      isCommentsLoading.value = false
+    }
+  }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useExpandedFormStore.ts:137]: ignore

File: packages/nocodb/src/meta/meta.service.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/meta/meta.service.ts:263]: ignore this

File: packages/nc-gui/components/smartsheet/expanded-form/RichComment.vue [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+        ?.chain()
+        .setTextSelection(docSize - 1)
+        .run()
+    }
+
+    ;(editor.value!.state as any).history$.prevRanges = null
+    ;(editor.value!.state as any).history$.done.eventCount = 0
+  }, 100)
+}
+
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/components/smartsheet/expanded-form/RichComment.vue:180]: ignore

File: packages/nc-gui/composables/useExpandedFormStore.ts [hunk 16]
```diff
@@ -87,24 +101,73 @@ const [useProvideExpandedFormStore, useExpandedFormStore] = useInjectionState((m
...
+      >
+
+      comments.value = res.map((comment) => {
+        const user = baseUsers.value.find((u) => u.id === comment.created_by)
+        return {
+          ...comment,
+          created_display_name: user?.display_name ?? (user?.email ?? '').split('@')[0],
+        }
+      })
+    } catch (e: any) {
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useExpandedFormStore.ts:132]: ignore

File: packages/nc-gui/composables/useExpandedFormStore.ts [hunk 17]
```diff
@@ -240,7 +306,7 @@
           addUndo({
             redo: {
               fn: async (id: string, data: Record<string, any>) => {
-                await $api.dbTableRow.update(NOCO, base.value.id as string, meta.value.id, encodeURIComponent(id), data)
+                await $api.dbTableRow.update(NOCO, base.value.id as string, meta.value.id!, encodeURIComponent(id), data)
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nc-gui/composables/useExpandedFormStore.ts:309]: The meta.value.id is checked if it is not null, before that segment

File: packages/nocodb/src/utils/acl.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/utils/acl.ts:214]: Not related to this pr

File: packages/nocodb/src/controllers/audits.controller.ts [hunk 18]
```diff
@@ -22,14 +20,12 @@
...
-  @HttpCode(200)
-  @Acl('commentRow')
-  async commentRow(@Req() req: Request) {
-    return await this.auditsService.commentRow({
-      user: (req as any).user,
-      body: req.body,
-    });
+  @Get(['/api/v1/db/meta/audits/', '/api/v2/meta/audits/'])
+  @Acl('auditList')
+  async auditListRow(@Req() req: Request) {
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/controllers/audits.controller.ts:25]: The decorator is required and correctly placed

File: packages/nocodb/src/services/comments.service.ts [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+      base: await Base.getByTitleOrId(model.base_id),
+      model: model,
+      user: param.user,
+      comment: param.body.comment,
+      rowId: param.body.row_id,
+      req: param.req,
+    });
+
+    return res;
+  }
```
[the remark this answers is not in the document]
Comment (DarkPhoenix2704) [reply packages/nocodb/src/services/comments.service.ts:44]: ignore the type suggestions

**Review by dstala — APPROVED** [review]

**Pull request merged**